include/future_queue.hpp

```diff
@@ -440,6 +442,77 @@ namespace cppext {
     };
 
+    inline void shared_state_base::free() {
+      // Reduce reference count but atomically keep the old reference counter. Note
+      // that the std::memory_order_relaxed refers to the access to the pointer not
+      // to the reference counter.
+      size_t oldCount = this->reference_count--;
+
+      // Determine whether we need to destroy the shared state depending on possible
+      // internal references.
+      bool executeDelete = false;
+
+      // Standard case: no continuation. If the last user is just destroying its
+      // reference we delete the shared state.
+      if(oldCount == 1 && !this->is_continuation_async && !this->is_continuation_deferred &&
+          !this->is_continuation_when_all) {
+        executeDelete = true;
+      }
+      // Deferred continuations (incl. when_all) have two internal use counts due to
+      // the two std::functions, so we need to remove those functions first.
+      else if(oldCount == 3 && (this->is_continuation_deferred || this->is_continuation_when_all)) {
+        this->continuation_process_deferred = {};
+        this->continuation_process_deferred_wait = {};
+        executeDelete = true;
+      }
+      // Async continuations have one internal use count inside their thread, so we
+      // need to terminate the thread first.
+      else if(oldCount == 2 && this->is_continuation_async) {
+        if(this->continuation_process_async.joinable()) {
+          // Signal termination to internal thread and wait until thread has been
+          // terminated
+          while(this->continuation_process_async_terminated == false) {
+            // Push a detail::TerminateInternalThread exception into the queue which
+            // the internal thread is potentially waiting on.
+            try {
+              throw detail::TerminateInternalThread();
+            }
+            catch(...) {
+              // Special case: the origin queue is a continuation itself (deferred
+              // or when_all) - we need to push the exception to the origin of the
+              // origin to actually reach the internal thread, since
+              // deferred/when_all continuations do not really use their own queue
+              if(this->continuation_origin.d->is_continuation_deferred ||
+                  this->continuation_origin.d->is_continuation_when_all) {
+                this->continuation_origin.d->continuation_origin.push_exception(std::current_exception());
+              }
+              // Standard case: just push the exception to the origin queue of the
+              // continuation
+              else {
+                this->continuation_origin.push_exception(std::current_exception());
+              }
+            } // end catch
+          }   // end while
+
+          this->continuation_process_async.join();
+        }
+        // end else branch 3.else
+        executeDelete = true;
+      } // end branch 3
```

Comment:
branch 3 has not been labeled as such. How about ``// end branch for async continuation`` (as described in line 468f)